reasoner/planner/ReasonerPlanPrinter.java

```diff
@@ -0,0 +1,181 @@
+        for (int i = 0; i < ruleMode.conjunctions.size(); i++) {
+            appendCallMode(ruleMode.conjunctions.get(i), ruleMode.mode, triggered, "");
+            if (ruleMode.conjunctions.size() - i > 1) appendBranchSeparator();
+        }
+        appendRuleModeSeparator();
+        printed.add(ruleMode);
+    }
+
+    private void appendCallMode(ResolvableConjunction conjunction, Set<Variable> mode, Set<RuleMode> triggered, String nesting) {
+        ReasonerPlanner.Plan plan = planner.getPlan(conjunction, mode);
+        Set<Variable> runningBounds = new HashSet<>(mode);
+
+        sb.append(nesting).append("Cost: ").append(plan.allCallsCost()).append("\n");
+        for (int i = 0; i < plan.plan().size(); i++) {
+            Resolvable<?> res = plan.plan().get(i);
+            if (res.isNegated()) {
+                final int index = i;
+                res.asNegated().disjunction().conjunctions().forEach(nestedConjunction -> {
+                    Set<Variable> negationMode = Collections.intersection(runningBounds, ReasonerPlanner.estimateableVariables(nestedConjunction.pattern().variables()));
+                    appendResolvableHeader(nesting, index, "NEG", negationMode);
+                    sb.append(nesting).append("{");
+                    sb.append(nesting).append("}\n");
+                    appendCallMode(nestedConjunction, negationMode, triggered, nesting + "\t");
+                });
+            } else {
+                Set<Variable> resolvableMode = Collections.intersection(runningBounds, ReasonerPlanner.estimateableVariables(res.variables()));
+                appendResolvable(res, i, resolvableMode, triggered, nesting);
+            }
+            if (!res.isNegated()) runningBounds.addAll(ReasonerPlanner.estimateableVariables(res.variables()));
+        }
+    }
+
+    private void appendResolvable(Resolvable<?> res, int index, Set<Variable> mode, Set<RuleMode> triggered, String nesting) {
+        if (res.isRetrievable()) {
+            appendResolvableHeader(nesting, index, "RET", mode);
+            appendResolvablePattern(nesting, res.asRetrievable().pattern());
+        } else if (res.isConcludable()) {
+            appendResolvableHeader(nesting, index, "CON", mode);
+            appendResolvablePattern(nesting, res.asConcludable().pattern());
+            triggeredRuleModes(res.asConcludable(), mode).forEach((ruleMode) -> {
+                triggered.add(ruleMode);
+                sb.append(nesting).append("\t\t- ").append(ruleMode.label).append("::").append(ruleMode.mode).append("\n");
+            });
+        }
+    }
+
+    private void appendRuleModeHeader(RuleMode ruleMode) {
+        sb.append("--------------------------------\t\t").append(ruleMode.label).append("::").append(ruleMode.mode).append("\t\t--------------------------------\n");
+    }
+
+    private void appendResolvableHeader(String nesting, int resolvableIndex, String resolvableType, Set<Variable> bounds) {
```

Comment:
is `resolvableIndex` the same as `planIndex`?

Reply:
@krishnangovindraj  this one too please

Reply:
It's the index of the resolvable in the plan (resolvable ordering) .